File: tree/Traversal/Preorder/recursive/java/PreorderTraversal.java
import java.util.*;

public class PreorderTraversal {
    public static List<Integer> PreorderTraversal(int[] leftChildren, int[] rightChildren, int root) {
        List<Integer> order = new ArrayList<>();
        traverse(root, leftChildren, rightChildren, order);
        return order;
    }

    private static void traverse(int node, int[] leftChildren, int[] rightChildren, List<Integer> order) {
        order.add(node);

        int leftChild = leftChildren[node];
        if (leftChild != -1) {
            traverse(leftChild, leftChildren, rightChildren, order);
        }

        int rightChild = rightChildren[node];
        if (rightChild != -1) {
            traverse(rightChild, leftChildren, rightChildren, order);
        }
    }
}

File: tree/Traversal/Preorder/recursive/rust/src/PreorderTraversal.rs
#[allow(non_snake_case)]
pub fn PreorderTraversal(left_children: &[i32], right_children: &[i32], root: usize) -> Vec<usize> {
    let mut order = Vec::new();

    fn traverse(node: usize, left_children: &[i32], right_children: &[i32], order: &mut Vec<usize>) {
        order.push(node);

        let left_child = left_children[node];
        if left_child != -1 {
            traverse(left_child as usize, left_children, right_children, order);
        }

        let right_child = right_children[node];
        if right_child != -1 {
            traverse(right_child as usize, left_children, right_children, order);
        }
    }

    traverse(root, left_children, right_children, &mut order);
    order
}
